refac: Create camera manager interface that will allow users to provide their own CameraManager

Savokr opened the pull request:
Cherry picked version of https://github.com/cognitedata/reveal/pull/1848 

Working version of CameraManager interface with all necessary fixes in internal code and requested ability to provide custom camera, etc.

larsmoa commented on viewer/packages/camera-manager/src/ComboControls.ts:
```diff
@@ -240,6 +243,10 @@ export class ComboControls extends EventDispatcher {
     this.triggerCameraChangeEvent();
   };
 
+  get cameraAdditionalRotation(): Quaternion {
+    return this._additionalRotation;
+  }
```
Dafuck

Savokr replied:
This is the key thing that got animation working. It is rotation that is applied after lookAt method was called. Value for it is set in `setCameraRotation` method and basically equal to the difference between desired rotation of the camera and the one we get after all manipulations in update function (lookAt method). So if we set rotation to the same value as the camera have right now, this additional rotation will become zero.

Savokr replied:
So, renamed to cameraRawRotation and changed the way it is handled a little bit. (As we describe on Wednesday 16.02).

larsmoa commented on viewer/packages/camera-manager/src/types.ts:
```diff
@@ -61,7 +61,89 @@ export interface RevealCameraControls extends THREE.EventDispatcher {
+   * @obvious
+   * @returns Camera's position in world space.
+   */
+  getCameraPosition: () => THREE.Vector3;
```
use plain function definitions, not lambda

Savokr commented on documentation/versioned_docs/version-3.x/examples/clipping.mdx:
```diff
@@ -116,14 +116,14 @@ controls.addEventListener('drag', () => {
 
 // When we are hovering over a sphere we want to disable viewer controls
 controls.addEventListener('hoveron', () => { 
-  viewer.cameraControlsEnabled = false;
+  viewer.cameraManager.cameraControlsEnabled = false;
   draggerMaterial.color.set('gray');
   viewer.requestRedraw();
 });
 
 // And turn them back on when we stop hovering over it
 controls.addEventListener('hoveroff', () => {
-  viewer.cameraControlsEnabled = true;
+  viewer.cameraManager.cameraControlsEnabled = true;
   draggerMaterial.color.set('white');
   viewer.requestRedraw();
 });
```
Oh, that's right

Savokr commented on examples/src/pages/Migration.tsx:
```diff
@@ -26,6 +26,7 @@ import { CameraUI } from '../utils/CameraUI';
 import { PointCloudUi } from '../utils/PointCloudUi';
 import { ModelUi } from '../utils/ModelUi';
 import { createSDKFromEnvironment } from '../utils/example-helpers';
+import { DefaultCameraManager } from '@cognite/reveal/packages/camera-manager';
```
Oh, it was automatically imported and I haven't corrected that

Savokr commented:
Fixed everything from comments and implemented setState method, so should be ready for another review. 
However, I thought about one problem that could arise when someone just want to pass custom `THREE.PerspectiveCamera` instance. The way our current `DefaultCameraManager` is implemented requires to pass `raycastFunction` in the constructor. It uses `Cognite3DViewer` data about the model to create proper raycast. So if someone just want to have custom camera without creating their own `CameraManager` it won't be possible with current implementation. 

My thoughts: we could either create an implementation-specific method in `DefaultCameraManager` to change camera in runtime so that people will be able to set it after `Cognite3DViewer` initialization or we can set `raycastFunction` in runtime and change the constructor of `DefaultCameraManager` a little bit.

Savokr commented on viewer/packages/camera-manager/src/CameraManager.ts:
```diff
@@ -71,34 +97,31 @@ export class CameraManager {
+    this._camera.position.x = 30;
+    this._camera.position.y = 10;
+    this._camera.position.z = 50;
+    this._camera.lookAt(new THREE.Vector3());
```
Removing it fails a few visual tests, so not sure what to do. We can either set it manually in visual tests or leave as is.

christjt commented:
Visual tests are failing, probably due to default camera position being removed (which is good). Just add it to the `Cognite3DTestViewer`, which should then apply to all visual tests.

larsmoa commented on viewer/core/src/public/migration/types.ts:
```diff
@@ -76,31 +77,14 @@ export interface Cognite3DViewerOptions {
+   * Camera manager instance that is used for controlling the camera.
+   * It is responsible for all manipulations that are done with the camera,
+   * including animations and modification of state. Also, gives ability
+   * to provide custom THREE.PerspectiveCamera instance to Cognite3DViewer.
```
```suggestion
   * to provide custom `THREE.PerspectiveCamera` instance to {@link Cognite3DViewer}.
```

larsmoa commented on viewer/packages/camera-manager/src/CameraManager.ts:
```diff
@@ -1,604 +1,58 @@
+import { CameraChangeData, CameraState } from './types';
+/**
+ * Camera manager interface.
+ * */
```
```suggestion
 */
```

larsmoa commented on viewer/packages/camera-manager/src/CameraManager.ts:
```diff
@@ -1,604 +1,58 @@
+export interface CameraManager {
   /**
-   * Reusable buffers used by functions in Cognite3dViewer to avoid allocations.
+   * Get THREE.PerspectiveCamera object for special cases. All manipulations of the camera should be done
+   * through other CameraManager methods.
+   * **/
```
```suggestion
   * Returns the camera used for rendering in {@link Cognite3DViewer}. 
   * Note that the camera will not be modified directly by Reveal. Implementations must trigger the `cameraChange`-event whenever the
   * camera changes.
   */
```

It's up to implementations if they want to modify the camera directly, whats important is that Reveal wont do that and that cameraChanged is triggered when it is modified.

larsmoa commented on viewer/packages/camera-manager/src/CameraManager.ts:
```diff
@@ -1,604 +1,58 @@
+  /**
+   * Set camera's state
+   *
+   * @param state Camera state, all fields are optional.
```
```
   * @param state Camera state, all fields are optional.
   * @param state.position ....
   * @param state.target ...
```

larsmoa commented on viewer/packages/camera-manager/src/CameraManager.ts:
```diff
@@ -1,604 +1,58 @@
   /**
-   * Calculates new camera target based on cursor position.
-   * @param event MouseEvent that contains pointer location data.
+   * Required for understadning of `cameraChanged` definition for other parts of Reveal
+   * @param event
+   * @param callback
```
```suggestion
   * Subscribe to changes to the camera. This is used by Reveal react to changes to the camera.
   * @param event
   * @param callback
```

Add some descriptions to `event` and `callback` too.

larsmoa commented on viewer/packages/camera-manager/src/CameraManager.ts:
```diff
@@ -1,604 +1,58 @@
+  on(event: 'cameraChange', callback: CameraChangeData): void;
   /**
-   * Removes controls event listeners if they are defined.
+   * Required for understadning of `cameraChanged` definition for other parts of Reveal
+   * @param event
+   * @param callback
```
Change to something about "unsubscribe to event", current description doesn't say what this function does - it says what the event does (which doesn't make sense in the unsubscribe function)

larsmoa commented on viewer/packages/camera-manager/src/CameraManager.ts:
```diff
@@ -1,604 +1,58 @@
-  }
+  fitCameraToBoundingBox(boundingBox: THREE.Box3, duration?: number, radiusFactor?: number): void;
   /**
-   * Method for setting up camera controls listeners and values inside current controls class.
+   * Updates internal state of camera manager.
+   * @param deltaTime
+   * @param boundingBox
    */
```
This doesn't tell anything to implementors about what they should do in their implementations. This description must be detailed and explain in detail what the function is expected to do and what the arguments are.

larsmoa commented on viewer/packages/camera-manager/src/DefaultCameraManager.ts:
```diff
@@ -0,0 +1,742 @@
+import { assertNever, EventTrigger, InputHandler, disposeOfAllEventListeners } from '@reveal/utilities';
+import range from 'lodash/range';
+
+export class DefaultCameraManager implements CameraManager {
```
Add description about how the default camera manage behaves and what it supports.

larsmoa commented on viewer/packages/camera-manager/src/DefaultCameraManager.ts:
```diff
@@ -0,0 +1,742 @@
+    return this._controls.enableKeyboardNavigation;
+  }
+
+  /**
+   * Allows to move camera with WASD or arrows keys.
+   */
+  enableKeyboardNavigation(): void {
+    this.keyboardNavigationEnabled = true;
+  }
+
+  /**z
+   * Disables camera movement by pressing WASD or arrows keys.
+   */
+  disableKeyboardNavigation(): void {
+    this.keyboardNavigationEnabled = false;
+  }
```
These are not necessary when we have the setter above, remove.

larsmoa commented on viewer/packages/camera-manager/tests/DefaultCameraManager.test.ts:
```diff
@@ -0,0 +1,43 @@
+import { DefaultCameraManager } from '../src/DefaultCameraManager';
+import { InputHandler } from '@reveal/utilities';
+
+describe('DefaultCameraManager', () => {
```
```suggestion
describe(DefaultCameraManager.name, () => {
```

This test could be extensively extended but I'm not sure it's worth it at the current stage.

Savokr replied:
Totally agree, I could add some tests in another PR.

larsmoa commented on viewer/packages/camera-manager/src/types.ts:
```diff
@@ -61,6 +61,20 @@ export interface RevealCameraControls extends THREE.EventDispatcher {
+   * Camera position in world space.
+   */
+  position?: THREE.Vector3;
+  /**
+   * Camera target in world space.
+   * **/
+  target?: THREE.Vector3;
+  /**
+   * Camera local rotation in quaternion form.
+   */
+  rotation?: THREE.Quaternion;
```
Not sure how to handle this, BUT target and rotation are conflicting. What would it take to get rid of `target` and make that a purely internal thing in `CameraManager`-implementations?

larsmoa replied:
Alternatively, should we define a precedence order - e.g. that rotation takes precedence over target when both are set?

Savokr replied:
That's how I implemented it in `DefaultCameraManager`. Target takes precedence when both are defined because I thought both of them are changing target, so explicit definition should have priority. Another option is to change rotation to only roll. That will remove the conflict as roll doesn't affect target-oriented rotation.

Savokr replied:
Now, setting rotation and target at the same time throws an error

larsmoa reviewed (changes requested):
Great leap forward and I love the super-simplistic nature of the `CameraManager`-interface which makes it easy to create simple implementations.

See a bunch of comments. I think the biggest question is around the `CameraState`-type and that `target` and `rotation` is conflicting, so we need to make a choice on how to handle that.

Savokr commented on viewer/packages/camera-manager/src/ComboControls.ts:
```diff
@@ -240,6 +245,10 @@ export class ComboControls extends EventDispatcher {
     this.triggerCameraChangeEvent();
   };
 
+  get cameraRawRotation(): Quaternion {
```
Added

larsmoa commented on viewer/packages/camera-manager/src/CameraManager.ts:
```diff
@@ -1,604 +1,68 @@
 
   /**
-   * Calculates new camera target based on cursor position.
-   * @param event MouseEvent that contains pointer location data.
+   * Get camera's state
    */
```
Add doc for each of the fields in `CameraState`, e.g.
```
@param state.position The camera position in world coordinates
```

Savokr replied:
Are you sure we want `@param` tag even on return? I added it as: 
`@returns Camera state: position, target and rotation.
`

Savokr replied:
Also, returning type describes how state looks, so I think it's ok to have such description

larsmoa commented on viewer/packages/camera-manager/src/CameraManager.ts:
```diff
@@ -1,604 +1,68 @@
-   * @param pixelX
-   * @param pixelY
+   * Updates internal state of camera manager. Expected to update visual state of the camera
+   * as well as it's near and far planes if needed. Called every time screen is refreshed.
```
`Called every time screen is refreshed.` <- this is wrong, right? It's actually called more often than that?

Savokr replied:
That's correct and one of the things that should be changed, I believe. Also, it's not every new rendered frame but every screen refresh with `requestAnimationFrame`.

larsmoa replied:
The doc should reflect what actually happens. I at least read this "every time the 3D view is re-rendered". A more precise version might be "Called in requestAnimationFrame-loop" and also emphasis that it's very important that the performance is high.

Savokr replied:
Changed description to: 
` /**
   * Updates internal state of camera manager. Expected to update visual state of the camera
   * as well as it's near and far planes if needed. Called in `requestAnimationFrame`-loop. 
   * Reveal performance affects frequency with which this method is called.
   * @param deltaTime Delta time since last update in seconds.
   * @param boundingBox Global bounding box of the model(s) and any custom objects added to the scene.
   */
  `

larsmoa commented on viewer/packages/camera-manager/src/ComboControls.ts:
```diff
@@ -94,6 +95,7 @@ export class ComboControls extends EventDispatcher {
   private readonly _spherical: Spherical = new Spherical();
   private _sphericalEnd: Spherical = new Spherical();
   private readonly _deltaTarget: Vector3 = new Vector3();
+  private readonly _rawCameraRotation: THREE.Quaternion = new Quaternion();
```
```suggestion
  private readonly _rawCameraRotation = new Quaternion();
```

larsmoa commented on viewer/packages/camera-manager/src/CameraManager.ts:
```diff
@@ -1,604 +1,75 @@
-  public automaticControlsSensitivity = true;
-
+import { CameraChangeData, CameraState } from './types';
+/**
```
```suggestion

/**
```

larsmoa commented on viewer/packages/camera-manager/src/CameraManager.ts:
```diff
@@ -1,604 +1,75 @@
-
+import { CameraChangeData, CameraState } from './types';
+/**
+ * Camera manager interface. It is responsible for all manipulations that are done with the camera,
+ * including animations and modification of state. Also, gives ability
+ * to provide custom `THREE.PerspectiveCamera` instance to {@link Cognite3DViewer}.
+ * For working implementation required to fire `cameraChange`-event when camera is changed.
+ * Default implementation is {@link DefaultCameraManager}.
+ */
```
```suggestion
 * Interface for manager responsible for all manipulations to the camera,
 * including animations and modification of state. Implementations are responsible for
 * providing a `THREE.PerspectiveCamera` instance to {@link Cognite3DViewer}. Implementations
 * must trigger the `cameraChange`-event when camera is changed.
 * The default implementation is {@link DefaultCameraManager}.
 */
```

larsmoa commented on viewer/packages/camera-manager/src/DefaultCameraManager.ts:
```diff
@@ -0,0 +1,686 @@
+import range from 'lodash/range';
+
+/**
+ * Default implementation of {@link CameraManager}. Uses {@link ComboControls} to control the camera.
```
Remove info about ComboControls now that we dont expose that anymore.

larsmoa commented on viewer/packages/camera-manager/src/CameraManager.ts:
```diff
@@ -1,604 +1,75 @@
-        assertNever(controlsOptions.mouseWheelAction);
-    }
-  }
+  on(event: 'cameraChange', callback: CameraChangeData): void;
```
Huh, is `CameraChangeData` the callback definition? Or the data in the event?

I think it should be renamed to `CameraChangedEvent`

Savokr replied:
Renamed

larsmoa commented on examples/src/pages/e2e/cad/DefaultCognite3DViewerTestPageV8.tsx:
```diff
@@ -2,15 +2,20 @@
   const modelUrl = 'primitives_v8';
 
-  return <Cognite3DTestViewer modelUrls={[modelUrl]} />;
+  return <Cognite3DTestViewer modelUrls={[modelUrl]} initializeCallback={(viewer:Cognite3DViewer) => {
+    viewer.cameraManager.setCameraState({position: new THREE.Vector3(30,10,50), 
+      target: new THREE.Vector3()});
+  }} />;
```
```suggestion
  return <Cognite3DTestViewer modelUrls={[modelUrl]} initializeCallback={(viewer:Cognite3DViewer) => {
    viewer.cameraManager.setCameraState({position: new THREE.Vector3(30,10,50), 
      target: new THREE.Vector3()});
  }} />;
```
Instead of fixing every single usage of `Cognite3DTestViewer`, just add this into the implementation of `Cognite3DTestViewer`.

larsmoa commented on viewer/packages/camera-manager/src/ComboControls.ts:
```diff
@@ -211,7 +213,10 @@ export class ComboControls extends EventDispatcher {
     _camera.position.setFromSpherical(_spherical).add(_target);
-    _camera.lookAt(this.lookAtViewTarget ? this._viewTarget : _target);
+
+    if (this._rawCameraRotation.equals(new Quaternion()))
```
Avoid `new` here - it's in the hot path and we should be very careful about allocations. Declare a variable outside the function or see if there is some other way of doing the same check without allocation.

larsmoa replied:
Also, add `{ ... }` for the if-else clauses

Savokr replied:
Create the `isIdentityQuaternion` function and check using it

larsmoa reviewed (changes requested):
See comments

larsmoa commented on viewer/packages/camera-manager/src/ComboControls.ts:
```diff
@@ -863,4 +879,8 @@ export class ComboControls extends EventDispatcher {
     _targetEnd.add(_panVector);
   };
+
+  private isIdentityQuaternion(q: THREE.Quaternion) {
+    return q.x === 0 && q.y === 0 && q.z === 0 && q.w === 1;
+  }
```
Cant this be used instead? https://threejs.org/docs/#api/en/math/Quaternion.identity

Savokr replied:
This is a method to make some quaternion an identity, not to check whether it is equal to identity.

larsmoa reviewed (approved):
Approved, but see single comment

larsmoa commented:
> This pull request **introduces 8 alerts** when merging [1df6187](https://github.com/cognitedata/reveal/commit/1df618755984a213f724757f1bd43c976094ef75) into [a2ba4ad](https://github.com/cognitedata/reveal/commit/a2ba4adfaa7a1a62c633b6a49109041b9b39d9c3) - [view on LGTM.com](https://lgtm.com/projects/g/cognitedata/reveal/rev/pr-f89797add37acb0102477171415fcb9fea7e72d9)
> 
> **new alerts:**
> 
> * 7 for Property access on null or undefined
> * 1 for Invocation of non-function

@Savokr
